### ISSUE TITLE
[SECURITY] AWS Next.js default caching behavior creates a DoS condition

### ISSUE THREAD

**Issue by K-Mistele** [sst/sst#4137]
# TL;DR
**Problem**: Several default AWS behaviors combine to create a Cache poisoning vulnerability that presents a Denial of Service issue.

**Solution**:
- Disable caching HTTP 5XX errors in the Cloudfront Cache policy for Next.js AWS apps
- Block the `X-HTTP-Method-Override` and `X-Method-Override` headers, or include them in the cloudfront cache key for a page. Currently only the Open Next `x-open-next-cache-key` is included.

# Explanation 
The default cloudfront policy for Next.js deployed to AWS with SST includes caching error pages. Additionally, the only HTTP header included in the cache key is `x-open-next-cache-key`.

_Why does this present a problem?_

### Cached Error Messages
* Sometimes, AWS API Gateways can be forced to return a 5XX error, e.g. a HTTP 502 error, by introducing malformed HTTP headers into the request
* For example, set a header line to be equal to `X-Amz-Blacklight=1:` 
* If the cache policy caches error messages, a maliciously crafted request's 502 error message can be cached for the request. Unless the header is part of the cache key (which is not possible, due to the numerous ways to set malformed headers), this error will be returned for all subsequent requests to the same route, regardless of whether or not they have the header, until the cache entry expires.
* A malicious actor can wait for a valid page in the application's cache to expire, then blast requests with the malformed header to cause a 5XX error to be returned and cached, and continue making this request to deny access to the page's contents. 
* Solution: do not cache 5XX errors.

### Method Override Headers
* Using the `X-Method-Override` or `X-HTTP-Method-Override` can be used in some cases to trigger a request of a different HTTP Verb than the request specifies. 
* For example, in many cases, a `GET` request with an `X-HTTP-Method-Override` header of `POST` will be treated as a `POST` request for the purposes of request processing by the app, but the response will be cached as though it were a `GET` request. 
* This can allow a cache to be poisoned with a `4XX` or `5XX` error, or an empty `HTTP 200` response triggered by what is effectively a `POST` request, to be cached for future responses to `GET` requests with the same route. 
* A malicious actor can repeatedly send such crafted requests to cause 4XX errors, 5XX errors, or an empty 200 response to be cached and returned for legitimate requests, effectively denying access to legitimate users. 
* _This behavior cannot be entirely prevented by preventing error messages from being cached_, since in some cases Next.js will return an empty 200 response instead of a `4XX` or `5XX` response when it receives such a request
* Therefore, the `X-HTTP-Method-Override` and `X-Method-Override` headers must be blocked from requests, or must be included in the cache key for items.

## Replication
We have run into this issue and seen reports to this effect for applications that are in scope for our bug bounty program, which we have replicated internally and determined that this is the source of the issue. 

Because these are cache-related errors, they can be hard to debug and reproduce.  However, steps to reproduce are provided below. They must be followed PRECISELY.

1. Create a next app with `create-next-app@latest`
2. Install SST into the app with `npx sst init`
3. Recommended: add a domain to the Next.js app in the SST configuration
4. Add a simple page at `/demo`, e.g. `export default async function Page() {console.log("demo");return <p>test</p>}`
5. Deploy the Application and _do not_ make any requests to the `/demo` route
6. Using Burp Suite or a similar tool, make the following Raw HTTP request, replacing the domain with the app's domain:

```
GET /demo HTTP/2
Host: yourdomain.yourwebsite.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X) Gecko Firefox
Accept: */*
Accept-Encoding: gzip, deflate, br
Content-Length: 2
X-Amz-BLACKLIGHT=1:
```

_this must be a raw request made over-the-wire e.g. with burp suite or netcat or something similar that allows you to edit a raw request_. Tools like `cURL` will often reject invalidly-formed headers before they send the request. 

You should receive a HTTP 502 error:
![Screenshot 2024-09-12 at 4 36 41 PM](https://github.com/user-attachments/assets/231cd31c-d382-4dd3-b97e-e5344bab4e52)

Visit `yourdomain.yourwebsite.com/demo` in your browser, and you will see that the HTTP 502 error has been cached. Usually, it will remain in the cache for about a minute. 

### "Gotchas"
- if you visit the page before making the malformed request, that inserts the "good" version of the response into the cache. When you make the request with the malformed header, a cache hit will happen and you will not successfully poison the cache because a cached value will be served back to your request
- You can use a query parameter like `?cache_buster=somethinggoeshere` to request the same page, but not get a cached version of the page. Subsequent requests with the same URL will be cached. This is a good way to test if your deployed app is vulnerable to this issue without actually risking poisoning the cache for a page users will see. For example, if you want to check if this works on your site's `/docs` route, you would make the initial request to `/docs?cache_buster=abc` and then visit that same URL (`/docs?cache_buster=abc`) in your browser to verify if the cache was poisoned, without affecting requests that are just `GET /docs`.
- This can be fundamentally difficult to test due to the nature of caching, so it may not work 100% of the time.
